sklearn/utils/metaestimators.py

```diff
@@ -80,53 +80,114 @@ def _validate_names(self, names):
             )
 
 
-class _IffHasAttrDescriptor:
+class _AvailableIfDescriptor:
     """Implements a conditional property using the descriptor protocol.
 
     Using this class to create a decorator will raise an ``AttributeError``
-    if none of the delegates (specified in ``delegate_names``) is an attribute
-    of the base object or the first found delegate does not have an attribute
-    ``attribute_name``.
-
-    This allows ducktyping of the decorated method based on
-    ``delegate.attribute_name``. Here ``delegate`` is the first item in
-    ``delegate_names`` for which ``hasattr(object, delegate) is True``.
+    if check(self) returns a falsey value. Note that if check raises an error
+    this will also result in hasattr returning false.
 
     See https://docs.python.org/3/howto/descriptor.html for an explanation of
     descriptors.
     """
 
-    def __init__(self, fn, delegate_names, attribute_name):
+    def __init__(self, fn, check, attribute_name):
         self.fn = fn
-        self.delegate_names = delegate_names
+        self.check = check
         self.attribute_name = attribute_name
 
         # update the docstring of the descriptor
         update_wrapper(self, fn)
 
-    def __get__(self, obj, type=None):
-        # raise an AttributeError if the attribute is not present on the object
+    def __get__(self, obj, owner=None):
         if obj is not None:
             # delegate only on instances, not the classes.
             # this is to allow access to the docstrings.
-            for delegate_name in self.delegate_names:
-                try:
-                    delegate = attrgetter(delegate_name)(obj)
-                except AttributeError:
-                    continue
-                else:
-                    getattr(delegate, self.attribute_name)
-                    break
-            else:
-                attrgetter(self.delegate_names[-1])(obj)
+            if not self.check(obj):
```

Comment:
What made you choose to not pass `self.attribute_name` here?

Reply:
I'm ambivalent to this change. It's mostly a question of what is more readable to someone who isn't otherwise familiar with `available_if`:

```py
    @available_if(base_estimator_has_method)
    def predict_proba(self, X):
```
or
```py
    @available_if(base_estimator_has("predict_proba"))
    def predict_proba(self, X):
```

Which do you think is clearer?

What about:
```py
    @available_if(lambda self, attr: hasattr(self.steps[-1][1], 'transform'))
    def fit_transform(self, X):
```
vs
```py
    @available_if(lambda self: hasattr(self.steps[-1][1], 'transform'))
    def fit_transform(self, X):
```
?

Reply:
Without knowing about `available_if`, I think **not** passing `self.attribute_name` is a little more readable.